sympy/integrals/transforms.py

```diff
@@ -1137,6 +1160,541 @@ def sbs(expr):
+        return f, a, c
+    else:
+        return f
+
+def _laplace_apply_rules(f, t, s, **hints):
+    doit = hints.pop('doit', False)
+    a = Wild('a', exclude=[t])
+    b = Wild('b', exclude=[t])
+    y = Wild('y')
+    z = Wild('z')
+    n = Wild('n', exclude=[t])
+    g = WildFunction('g', nargs=1)
+    k, func = f.as_independent(t, as_Add=False)
+    # if not any(func.has(t) for func in f.atoms(AppliedUndef)):
+    simple_rules = _laplace_build_rules(t, s)
+    for t_dom, s_dom, check, plane, prep in simple_rules:
+        ma = prep(func).match(t_dom)
+        if ma:
+            debug('_laplace_apply_rules match:')
+            debug('      f:    %s'%(func,))
+            debug('      rule: %s o---o %s'%(t_dom, s_dom))
+            try:
+                debug('      try   %s'%(check,))
+                c = check.xreplace(ma)
+                debug('      check %s -> %s'%(check, c))
+                if c==True:
+                    debug(' ---> go')
+                    return _laplace_cr(k*s_dom.xreplace(ma),
+                                plane.xreplace(ma), S.true, **hints)
+            except Exception:
+                debug('_laplace_apply_rules did not match.')
+    # else:
+    if f.has(DiracDelta):
+        raise IntegralTransformError(
+                    'Laplace', None, 'No rules could be applied.')
+    ma1 = func.match(g)
+    if ma1:
+        arg = ma1[g].args[0].collect(t)
+        ma2 = arg.match(b*t)
+        if ma2 and ma2[b]>0:
+            debug('_laplace_apply_rules match:')
+            debug('      f:    %s ( %s, %s )'%(f, ma1, ma2))
+            debug('      rule: amplitude and time scaling (1.1, 1.2)')
+            if ma2[b]==1:
+                if doit==True and not any(func.has(t) for func
+                                          in ma1[g].atoms(AppliedUndef)):
+                    return k*_laplace_transform(ma1[g].func(t), t, s, **hints)
+                else:
+                    return k*LaplaceTransform(ma1[g].func(t), t, s, **hints)
+            else:
+                L = _laplace_apply_rules(ma1[g].func(t), t, s/ma2[b], **hints)
+                try:
+                    r, p, c = L
+                    return (k/ma2[b]*r, p, c)
+                except TypeError:
+                    return k/ma2[b]*L
+    ma1 = func.match(Heaviside(y)*g)
+    if ma1:
+        ma2 = ma1[y].match(t-a)
+        ma3 = ma1[g].args[0].collect(t).match(t-b)
+        if ma2 and ma2[a]>0 and ma3 and ma2[a]==ma3[b]:
+            debug('_laplace_apply_rules match:')
+            debug('      f:    %s ( %s, %s, %s )'%(f, ma1, ma2, ma3))
+            debug('      rule: time shift (1.3)')
+            L = _laplace_apply_rules(ma1[g].func(t), t, s, **hints)
+            try:
+                r, p, c = L
+                return (k*exp(-ma2[a]*s)*r, p, c)
+            except TypeError:
+                return k*exp(-ma2[a]*s)*L
+    ma1 = func.match(exp(y)*z)
+    if ma1:
+        ma2 = ma1[y].collect(t).match(a*t)
+        if ma2:
```

Comment:
The code in this section seems to be a bit repetitive. Can these rules be abstracted out somehow or at least made into named functions or something?

Reply:
Maybe. I knew that you would ask this, and thought about it, but did not find a good way. Since there are only five similar rules (exp, sin, cos, sinh, cosh) and more are not expected to come, and there are small differences in what they return, I decided against abstracting it.

Reply:
They could at least be in named functions so that the high-level logic in this function is easier to parse. Something like:
```python
def _exp_rule(...):
    ...

def _laplace_transform(func):
    rules = [_exp_rule, _sin_rule, ...]
    for rule in rules:
        result = rule(func)
        if result is not None:
            return result
```